### PR TITLE
Reverify programs that are extended using ExtendProgram

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
The programs extended via `ExtendProgram` instruction need re-verification, as the updated program may return different verification result than the original program.

#### Summary of Changes
- Compile and verify the program with the extended size
- Add a cooldown of 1 slot to prevent repeated extension of the program in the same slot
- Add a new unit test for the cooldown
- Update existing tests to deploy a program and extend it

Fixes #

### CODE REVIEW

**Comment by Lichtso** [comment]
Maybe we should actually perform the verification / compilation inside the instruction (like we do for upgrades) so that it can fail the TX if the verification fails.

File: programs/bpf_loader/src/lib.rs [hunk 1]
```diff
@@ -1403,6 +1406,29 @@ fn process_loader_upgradeable_instruction(
                 .try_borrow_instruction_account(transaction_context, PROGRAM_DATA_ACCOUNT_INDEX)?;
             programdata_account.set_data_length(new_len)?;
 
+            let slot = invoke_context
+                .get_sysvar_cache()
+                .get_clock()
+                .map(|clock| clock.slot)
+                .unwrap_or(invoke_context.programs_modified_by_tx.slot());
```
Comment (Lichtso) [inline programs/bpf_loader/src/lib.rs:1413]: I think this should fail if the cache does not contain the sysvar, just use `?`instead of `.unwrap_or()`.

**Review by Lichtso — DISMISSED** [review]

**Review by Lichtso — APPROVED** [review]

**Comment by jstarry** [comment]
Note that since this change now updates the program data clock slot, it's no longer possible to extend and update a program in the same slot (or transaction for that matter). I don't think it's a great dev experience that two separate transactions will need to be created for program upgrades but the extend program instruction is still an improvement over the current dev experience so I think it's ok for now.

File: programs/bpf_loader/src/lib.rs [hunk 2]
```diff
@@ -1403,6 +1417,30 @@ fn process_loader_upgradeable_instruction(
                 .try_borrow_instruction_account(transaction_context, PROGRAM_DATA_ACCOUNT_INDEX)?;
             programdata_account.set_data_length(new_len)?;
 
+            let programdata_data_offset = UpgradeableLoaderState::size_of_programdata_metadata();
+
+            deploy_program!(
+                invoke_context,
+                program_key,
+                &program_id,
+                UpgradeableLoaderState::size_of_program().saturating_add(new_len),
```
Comment (jstarry) [inline programs/bpf_loader/src/lib.rs:1426]: I think `new_len` should already be the full account size, why is `size_of_program` added here as well?

Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:1426]: I think you are right, this is a bug, probably copy pasta.
A fix should be backported before the feature is activated.

@pgarg66 Thoughts?

**Pull request merged**